## Data Wrangling Twitter Dogs

The aim of this project is to show proficiency at:

1. Gathering multiple sources of data  - via API and prepared flatfile storage
2. Assessing the data for issues:
3. Cleaning the data
4. Visualizing the data

The tools used to do this are: Python3, Jupyter
Python libraries used: Pandas, Requests, Tweepy, Matplotlib

The project is split into 4 stages - Gather Assess Clean and Visualise

## Section 1: GATHER THE DATA

Gather each of the three pieces of data as described below in a Jupyter Notebook titled wrangle_act.ipynb:

    The WeRateDogs Twitter archive. I am giving this file to you, so imagine it as a file on hand. Download this file manually by clicking the following link: twitter_archive_enhanced.csv - DONE

    The tweet image predictions, i.e., what breed of dog (or other object, animal, etc.) is present in each tweet according to a neural network. This file (image_predictions.tsv) is hosted on Udacity's servers and should be downloaded programmatically using the Requests library and the following URL: https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv

    Each tweet's retweet count and favorite ("like") count at minimum, and any additional data you find interesting. Using the tweet IDs in the WeRateDogs Twitter archive, query the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file called tweet_json.txt file. Each tweet's JSON data should be written to its own line. Then read this .txt file line by line into a pandas DataFrame with (at minimum) tweet ID, retweet count, and favorite count. Note: do not include your Twitter API keys, secrets, and tokens in your project submission.


In [1]:
#gather the twitter archive data
import pandas as pd
arch = "twitter-archive-enhanced.csv"
base_df = pd.read_csv(arch)
base_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [2]:
#gather the cloudfront image prediction data
import requests
from io import StringIO

r = requests.get("https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv")
type(r.text)
data = StringIO(r.text)
img_pred_df = pd.read_csv(data, sep="\t")
img_pred_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

#pull the additional columns(retweet and fav count at min) from tweepy#
import tweepy as twp

#NOTE need to be careful not to mess up the twitter API - 15min blocks.
import tweepy

consumer_key = ''
consumer_secret = ''
access_token = ""
access_secret = '

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

test_tweet = "666020888022790149"

tweet = api.get_status(test_tweet, tweet_mode="extended")

print(tweet._json.keys())

#COMMENT OUT AND DON'T RERUN
#THIS CODE SCRAPES TWITTER
#THIS DATA IS SAVED IN twiter_scraped.json
import json
#big dictionary to store all the data - should come in handy as I suspect based on the test tweet
#that there's some sort of mismatch in the tweet ID's 
big_json = {}
failed_scrapes = []

#now run the thing
for tweet_id in base_df.tweet_id:
    try:
        tweet = api.get_status(tweet_id, tweet_mode="extended")
        print("Tweet no. {} retrieved successfully!".format(tweet_id))
        big_json[tweet_id]= tweet._json
    except tweepy.TweepError as e:
        print("Problem retrieving tweet no. {}".format(tweet_id))
        failed_scrapes.append(tweet_id)





#showing how i stored the data, don't rerun
with open("twitter_scraped.json", "w") as file:
    file.write(js)

In [27]:
# Open previously scraped  twitter data.
with open("twitter_scraped.json", "r") as file:
    data2 = json.load(file)

data2.keys()

dict_keys(['892420643555336193', '892177421306343426', '891815181378084864', '891689557279858688', '891327558926688256', '891087950875897856', '890971913173991426', '890729181411237888', '890609185150312448', '890240255349198849', '890006608113172480', '889880896479866881', '889665388333682689', '889638837579907072', '889531135344209921', '889278841981685760', '888917238123831296', '888804989199671297', '888554962724278272', '888078434458587136', '887705289381826560', '887517139158093824', '887473957103951883', '887343217045368832', '887101392804085760', '886983233522544640', '886736880519319552', '886680336477933568', '886366144734445568', '886267009285017600', '886258384151887873', '886054160059072513', '885984800019947520', '885528943205470208', '885518971528720385', '885311592912609280', '885167619883638784', '884925521741709313', '884876753390489601', '884562892145688576', '884441805382717440', '884247878851493888', '884162670584377345', '883838122936631299', '883482846933004288',

In [29]:
data_df = pd.DataFrame.from_dict(data2, orient="index")
data_df.head()

created_at                  id  \
892420643555336193  Tue Aug 01 16:23:56 +0000 2017  892420643555336193   
892177421306343426  Tue Aug 01 00:17:27 +0000 2017  892177421306343426   
891815181378084864  Mon Jul 31 00:18:03 +0000 2017  891815181378084864   
891689557279858688  Sun Jul 30 15:58:51 +0000 2017  891689557279858688   
891327558926688256  Sat Jul 29 16:00:24 +0000 2017  891327558926688256   

                                id_str  \
892420643555336193  892420643555336193   
892177421306343426  892177421306343426   
891815181378084864  891815181378084864   
891689557279858688  891689557279858688   
891327558926688256  891327558926688256   

                                                            full_text  \
892420643555336193  This is Phineas. He's a mystical boy. Only eve...   
892177421306343426  This is Tilly. She's just checking pup on you....   
891815181378084864  This is Archie. He is a rare Norwegian Pouncin...   
891689557279858688  This is Darla. She commenced a snooze mid meal...   
891327558926688256  This is Franklin. He would like you to stop ca...   

                    truncated display_text_range  \
892420643555336193      False            [0, 85]   
892177421306343426      False           [0, 138]   
891815181378084864      False           [0, 121]   
891689557279858688      False            [0, 79]   
891327558926688256      False           [0, 138]   

                                                             entities  \
892420643555336193  {'hashtags': [], 'symbols': [], 'user_mentions...   
892177421306343426  {'hashtags': [], 'symbols': [], 'user_mentions...   
891815181378084864  {'hashtags': [], 'symbols': [], 'user_mentions...   
891689557279858688  {'hashtags': [], 'symbols': [], 'user_mentions...   
891327558926688256  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                                    extended_entities  \
892420643555336193  {'media': [{'id': 892420639486877696, 'id_str'...   
892177421306343426  {'media': [{'id': 892177413194625024, 'id_str'...   
891815181378084864  {'media': [{'id': 891815175371796480, 'id_str'...   
891689557279858688  {'media': [{'id': 891689552724799489, 'id_str'...   
891327558926688256  {'media': [{'id': 891327551943041024, 'id_str'...   

                                                               source  \
892420643555336193  <a href="http://twitter.com/download/iphone" r...   
892177421306343426  <a href="http://twitter.com/download/iphone" r...   
891815181378084864  <a href="http://twitter.com/download/iphone" r...   
891689557279858688  <a href="http://twitter.com/download/iphone" r...   
891327558926688256  <a href="http://twitter.com/download/iphone" r...   

                    in_reply_to_status_id  ... favorited  retweeted  \
892420643555336193                    NaN  ...     False      False   
892177421306343426                    NaN  ...     False      False   
891815181378084864                    NaN  ...     False      False   
891689557279858688                    NaN  ...     False      False   
891327558926688256                    NaN  ...     False      False   

                   possibly_sensitive possibly_sensitive_appealable lang  \
892420643555336193              False                         False   en   
892177421306343426              False                         False   en   
891815181378084864              False                         False   en   
891689557279858688              False                         False   en   
891327558926688256              False                         False   en   

                   retweeted_status quoted_status_id quoted_status_id_str  \
892420643555336193              NaN              NaN                  NaN   
892177421306343426              NaN              NaN                  NaN   
891815181378084864              NaN              NaN                  NaN   
891689557279858688              NaN              NaN                  NaN   
8913

In [24]:
base_df.loc[base_df.tweet_id == 870374049280663552]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
115  870374049280663552                    NaN                  NaN   

                     timestamp  \
115  2017-06-01 20:18:38 +0000   

                                                source  \
115  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
115  This is Zoey. She really likes the planet. Wou...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
115                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
115  https://twitter.com/dog_rates/status/870374049...                13   

     rating_denominator  name doggo floofer pupper puppo  
115                  10  Zoey  None    None   None  None

In [23]:
base_df.tweet_id.loc[]

0       892420643555336193
1       892177421306343426
2       891815181378084864
3       891689557279858688
4       891327558926688256
               ...        
2351    666049248165822465
2352    666044226329800704
2353    666033412701032449
2354    666029285002620928
2355    666020888022790149
Name: tweet_id, Length: 2356, dtype: int64

## Gathering Data 1
"Each piece of data is imported into a separate padnas DataFrame at first"

base_df = contains the data provided by Udacity via TSV /n

data_df = contains all the data scraped from the twitter API. This includes ALL subfields, as I want the option of exploring this later on. /n

img_pred_df = contains the image predictions from udacity's cloudfront via HTTP Get request

In [31]:
base_df.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

In [34]:
#bit of cleaning required to extract tweet_id from index to make a column
data_df.head(2)
data_df.reset_index(inplace=True)
data_df = data_df.rename(columns = {"index": "tweet_id"})

In [35]:
data_df.head(2)

tweet_id                      created_at                  id  \
0  892420643555336193  Tue Aug 01 16:23:56 +0000 2017  892420643555336193   
1  892177421306343426  Tue Aug 01 00:17:27 +0000 2017  892177421306343426   

               id_str                                          full_text  \
0  892420643555336193  This is Phineas. He's a mystical boy. Only eve...   
1  892177421306343426  This is Tilly. She's just checking pup on you....   

   truncated display_text_range  \
0      False            [0, 85]   
1      False           [0, 138]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str'...   
1  {'media': [{'id': 892177413194625024, 'id_str'...   

                                              source  ...  favorited  \
0  <a href="http://twitter.com/download/iphone" r...  ...      False   
1  <a href="http://twitter.com/download/iphone" r...  ...      False   

  retweeted  possibly_sensitive possibly_sensitive_appealable lang  \
0     False               False                         False   en   
1     False               False                         False   en   

  retweeted_status quoted_status_id quoted_status_id_str  \
0              NaN              NaN                  NaN   
1              NaN              NaN                  NaN   

  quoted_status_permalink quoted_status  
0                     NaN           NaN  
1                     NaN           NaN  

[2 rows x 33 columns]

In [33]:
img_pred_df.head(2)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True

## Decide what extra data to keep from the API - no need to have everything

In [36]:
data_df.columns

Index(['tweet_id', 'created_at', 'id', 'id_str', 'full_text', 'truncated',
       'display_text_range', 'entities', 'extended_entities', 'source',
       'in_reply_to_status_id', 'in_reply_to_status_id_str',
       'in_reply_to_user_id', 'in_reply_to_user_id_str',
       'in_reply_to_screen_name', 'user', 'geo', 'coordinates', 'place',
       'contributors', 'is_quote_status', 'retweet_count', 'favorite_count',
       'favorited', 'retweeted', 'possibly_sensitive',
       'possibly_sensitive_appealable', 'lang', 'retweeted_status',
       'quoted_status_id', 'quoted_status_id_str', 'quoted_status_permalink',
       'quoted_status'],
      dtype='object')

In [ ]:
["tweet_id", "favorite_count", "retweet_count", "favorited", "possibly_sensitive"]

In [40]:
data_r = data_df[["tweet_id", "full_text", "favorite_count", "retweet_count", "favorited", "possibly_sensitive"]]

In [41]:
data_r.head()

tweet_id                                          full_text  \
0  892420643555336193  This is Phineas. He's a mystical boy. Only eve...   
1  892177421306343426  This is Tilly. She's just checking pup on you....   
2  891815181378084864  This is Archie. He is a rare Norwegian Pouncin...   
3  891689557279858688  This is Darla. She commenced a snooze mid meal...   
4  891327558926688256  This is Franklin. He would like you to stop ca...   

   favorite_count  retweet_count  favorited possibly_sensitive  
0           35458           7487      False              False  
1           30695           5557      False              False  
2           23088           3680      False              False  
3           38741           7664      False              False  
4           37026           8265      False              False

# Assessing data
Variable reminder:

    data_r = API data
    
    img_pred_df = cloudfront udacity img classifier predictions
    
    base_df = udacity TSV data

## data_r - Visual

In [43]:
#Start with thte API Data - visual inspection
data_r.head(5)

tweet_id                                          full_text  \
0  892420643555336193  This is Phineas. He's a mystical boy. Only eve...   
1  892177421306343426  This is Tilly. She's just checking pup on you....   
2  891815181378084864  This is Archie. He is a rare Norwegian Pouncin...   
3  891689557279858688  This is Darla. She commenced a snooze mid meal...   
4  891327558926688256  This is Franklin. He would like you to stop ca...   

   favorite_count  retweet_count  favorited possibly_sensitive  
0           35458           7487      False              False  
1           30695           5557      False              False  
2           23088           3680      False              False  
3           38741           7664      False              False  
4           37026           8265      False              False

No obvious issues or insights with a quick visual inspection. In order to find out anything useful we'll need to investigate programmatically

## data_r - programmatic

In [45]:
data_r.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 6 columns):
tweet_id              2331 non-null object
full_text             2331 non-null object
favorite_count        2331 non-null int64
retweet_count         2331 non-null int64
favorited             2331 non-null bool
possibly_sensitive    2197 non-null object
dtypes: bool(1), int64(2), object(3)
memory usage: 93.5+ KB


In [83]:
# returns nothing, so no duplicate records to worry about
assert [data_r.duplicated() == False]

In [88]:
data_r.full_text

0       This is Phineas. He's a mystical boy. Only eve...
1       This is Tilly. She's just checking pup on you....
2       This is Archie. He is a rare Norwegian Pouncin...
3       This is Darla. She commenced a snooze mid meal...
4       This is Franklin. He would like you to stop ca...
                              ...                        
2326    Here we have a 1949 1st generation vulpix. Enj...
2327    This is a purebred Piers Morgan. Loves to Netf...
2328    Here is a very happy pup. Big fan of well-main...
2329    This is a western brown Mitsubishi terrier. Up...
2330    Here we have a Japanese Irish Setter. Lost eye...
Name: full_text, Length: 2331, dtype: object

Overall: 2331 records - compare with others.

tweet_id - currently a string. at first glance this should be an int, but this is an ordinal variable - we aren't going to be adding or subtracting these. It should stay as a string.

full_text - stays as a string
favorite_count - stays as an int
retweet_count - stays as an int
favorited - Boolean = correct, keep as-is
possibly_sensitive = should be fixed- there's some missing data. also it should be a boolean.

actions required:

1. Change possibly_sensitive column from string to bool. Decide what to do with nonvalues



The author's tweets are themselves fairly unreliable both in structure and in content. For example, Phineas is not "a mystical boy" -he is a specific breed of dog. The author's penchant for lies and inaccuracies means that we can't rely on this column to accurately label our dog. Similarly, vulpix is a pokemon, Mitsubishi is a japanese car, and Piers Morgan is a reactionary greasebag tv presenter - therefore we can't really use the Regex parsing techniques outlined in the course safely on this column to extract reliable information from this column.



In [57]:
#are there any duplioates?
len(data_r.tweet_id.unique())
#confirmed, no duplicates

2331

In [ ]:
data_r.possibly_sensitive.value_counts(dropna=False)

There are 134 missing records for this column - this doesn't mean that 
those records are possibly_sensitive, just that they are not definitely not
possibly_sensitive.

Again, absent a way of verifying them, I can't tell if some NaN's should be True. At this stage
I'm going to drop this column from my analysis as it doesn't tell us anything.

## img_pred_df Visual

In [51]:
img_pred_df.head(5)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

## img_pred_df Programmatic

In [52]:
img_pred_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


first off, we have fewer than 2331 values so not every tweet in the udacity TSV received a 

In [60]:
len(img_pred_df.tweet_id.unique())#
#at least all of these are unique

2075

In [82]:
# returns nothing, so no duplicate records to worry about
assert [img_pred_df.duplicated() == False]

Analysis and Issues :


tweet_id    2075 non-null int64 - fine, should change to a string though for consistency and joining.
jpg_url     2075 non-null object - don't really need for analysis
img_num     2075 non-null int64 - not really clear what this means?
p1          2075 non-null object - looks fine, however will need to check if the algo is using the same set of breeds as the labelled data.
p1_conf     2075 non-null float64 - looks fine
p1_dog      2075 non-null bool - looks fine

#will probably just drop these columns - only care about the top predictions

p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool

In [61]:
img_pred_df.p1.value_counts()

golden_retriever      150
Labrador_retriever    100
Pembroke               89
Chihuahua              83
pug                    57
                     ... 
carousel                1
platypus                1
cowboy_boot             1
pillow                  1
pot                     1
Name: p1, Length: 378, dtype: int64

In [63]:
img_pred_df.p1_dog.value_counts()

True     1532
False     543
Name: p1_dog, dtype: int64

clearly the algorithm is detecting some other objects than dogs - this isn't a data quality issue however.
I would just bear this in mind when combining the prediction with the main table.
It's arguable that to tidy this data, the rows should be individual predictions, rather than individual tweets - but I'm sidestepping this by only caring about the prediction that the algorithm is most confident about, for each tweet.



## base_df Visual

In [78]:
base_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

### notes:

From the project page: "I extracted this data programmatically, but I didn't do a very good job. The ratings probably aren't all correct. Same goes for the dog names and probably dog stages (see below for more information on these) too. You'll need to assess and clean these columns if you want to use them for analysis and visualization."


looks like in_reply_to columns may be sparsely populated - not an issue.
Timestamp formatting looks consistent but should be checked programmatically
Same for source

retweeted_status columns look sparse

expanded urls look fine

rating numerator and rating denominator should probably be combined into Rating for Tidy

Might be possible to validate this against FULLTEXT - the problem is the author's natural language might cause them to use the pattern "/10" in an awkward place, but there's grounds for a regex there.

name - again , could be regex validated but subject to same caveats about accuracy

doggo/floofer/pupper/puppo - again, author's use of language and reference to subjects outside the dog of concern in the tweet means we can't just find one of these patterns in the full text and assume that this is the category for this dog.IE what I'm saying is that the Fulltext column can't be used as a source of truth for this. We would probably have to have an extra credit student project where instead of dog breeds the classifier is trying to predict Doggo_stage from the related image. I have too much self respect to carry out such a project.

However theres no reason for these columns to be one hot encoded - consolidate / melt them into a single Doggo_Stage categorical variable with 4 levels - puppo, pupper, floofer, doggo

## base_df Programmatic

In [79]:
base_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

tweet_id                      2356 non-null int64 - can be object, we aren't doing maths here
in_reply_to_status_id         78 non-null float64 -as above. seems like these are replies?
in_reply_to_user_id           78 non-null float64 - as above.
timestamp                     2356 non-null object - should be datetime
source                        2356 non-null object  - no issues
text                          2356 non-null object - no issues
retweeted_status_id           181 non-null float64 - these are retweets?
retweeted_status_user_id      181 non-null float64 - n oissues
retweeted_status_timestamp    181 non-null object - no issues
expanded_urls                 2297 non-null object - no issues, some missing
rating_numerator              2356 non-null int64 
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object

In [80]:
# returns nothing, so no duplicate records to worry about
assert [base_df.duplicated() == False]

#### Addressing named columns in the project details

#### Ratings Columns


In [84]:
base_df.rating_numerator.value_counts() # weirdly no 16 out of 10s

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [89]:
#hehe many wow much drugreference
base_df[base_df.rating_numerator == 420]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
188   855862651834028034           8.558616e+17          194351775.0   
2074  670842764863651840                    NaN                  NaN   

                      timestamp  \
188   2017-04-22 19:15:32 +0000   
2074  2015-11-29 05:52:33 +0000   

                                                 source  \
188   <a href="http://twitter.com/download/iphone" r...   
2074  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
188   @dhmontgomery We also gave snoop dogg a 420/10...                  NaN   
2074  After so many requests... here you go.\n\nGood...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
188                        NaN                        NaN   
2074                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
188                                                 NaN               420   
2074  https://twitter.com/dog_rates/status/670842764...               420   

      rating_denominator  name doggo floofer pupper puppo  
188                   10  None  None    None   None  None  
2074                  10  None  None    None   None  None

In [90]:
base_df[base_df.rating_numerator == 666]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
189  855860136149123072           8.558585e+17           13615722.0   

                     timestamp  \
189  2017-04-22 19:05:32 +0000   

                                                source  \
189  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
189  @s8n You tried very hard to portray this good ...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp expanded_urls  \
189                       NaN                        NaN           NaN   

     rating_numerator  rating_denominator  name doggo floofer pupper puppo  
189               666                  10  None  None    None   None  None

Ok so another data quality issue is that the author uses the ratings column to make jokes / express things that relate to the image. Therefore we need to apply some sort of sense to this column - Based on how the project details described how this column works, I'm thinking of using a method (clip?) to truncate the possible values in this column to between 1 and 20, with Nan if it falls outside that range.

In [91]:
# Ratings columns 2 - denominator
base_df.rating_denominator.value_counts() #

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [93]:
base_df[base_df.rating_denominator == 120]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1779  677716515794329600                    NaN                  NaN   

                      timestamp  \
1779  2015-12-18 05:06:23 +0000   

                                                 source  \
1779  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1779                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1779  https://twitter.com/dog_rates/status/677716515...               144   

      rating_denominator  name doggo floofer pupper puppo  
1779                 120  None  None    None   None  None

Data quality issue here - based on the project details this column should just be ten. Again the author is using different denominators to be fun and expressive - this is not allowed. 

DQ ISSUE - just noticed somethign in the description - the author is actually rating multiple dogs in a single picture! This would explain why the denominator is so high - and may also explain some of the high numerators too.

This means i need to make a call on how to 1. Identify and 2. Treat these cases.

1. Identifying - I can't just write a rule that says "ratings over 20 must be multiples" - because there's 420 and 666 and various other joke ratings.

2. Treating - We can't keep suspected multiple dog images in the dataset without breaking the rules of Tidy Data - each row has to refer to a single observation, which in our case is assumed to be a single tweet about a single dog. There is no way of breaking a single tweet into multiple dogs. Moreover the img_pred_df doesn't tell us which dog in the image is being predicted.

3. Conclusion - given that both identifying and treating these cases would have to be done on a case by case basis for want of a clear rule based approach that resolves these issues, I propose to DROP THESE VALUES FROM THE DATA ENTIRELY, ie:

if numerator is outside range 1-20:  DROP

if denominator is not 10: DROP

Data Tidiness issue here - if this whole column is just 10, let's drop it entirely and rename the numerator column to rating_out_of_10.


# Intermission - clunkiness issue

Having identified several Data Quality and Tidiness issues above, the next step in the analysis before cleaning is to bring the datasets together. This means that I don't have to remember 3 variable names.

Once this is done I can get to cleaning.

First I'll left join base_df <- img_pred_df

Then I'll left join double_df <- data_r.

I'll do this all using tweet_id, as that's the common key between the three.


In [94]:
merged1 = pd.merge(base_df, img_pred_df, how="left", left_on=["tweet_id"], right_on=["tweet_id"])
merged1.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  ...  img_num  \
0  https://twitter.com/dog_rates/status/892420643...  ...      1.0   
1  https://twitter.com/dog_rates/status/892177421...  ...      1.0   
2  https://twitter.com/dog_rates/status/891815181...  ...      1.0   
3  https://twitter.com/dog_rates/status/891689557...  ...      1.0   
4  https://twitter.com/dog_rates/status/891327558...  ...      2.0   

            p1   p1_conf p1_dog                  p2   p2_conf p2_dog  \
0       orange  0.097049  False               bagel  0.085851  False   
1    Chihuahua  0.323581   True            Pekinese  0.090647   True   
2    Chihuahua  0.716012   True            malamute  0.078253   True   
3  paper_towel  0.170278  False  Labrador_retriever  0.168086   True   
4       basset  0.555712   True    English_springer  0.225770   True   

                            p3   p3_conf p3_dog  
0                       banana  0.076110  False  
1                     papillon  0.068957   True  
2                       kelpie  0.031379   True  
3                      spatula  0.040836  False  
4  German_short-haired_pointer  0.175219   True  

[5 rows x 28 columns]

In [100]:
#need to change data_r.tweet_id from object to int so can merge

data_r.tweet_id = data_r.tweet_id.astype("int64")
data_r.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 6 columns):
tweet_id              2331 non-null int64
full_text             2331 non-null object
favorite_count        2331 non-null int64
retweet_count         2331 non-null int64
favorited             2331 non-null bool
possibly_sensitive    2197 non-null object
dtypes: bool(1), int64(3), object(2)
memory usage: 93.5+ KB


C:\Users\44774\Anaconda3\envs\tf2\lib\site-packages\pandas\core\generic.py:5208: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [101]:
merged2 = pd.merge(merged1, data_r, how="left", left_on=["tweet_id"], right_on=["tweet_id"])
merged2.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  ...   p2_conf  p2_dog  \
0  https://twitter.com/dog_rates/status/892420643...  ...  0.085851   False   
1  https://twitter.com/dog_rates/status/892177421...  ...  0.090647    True   
2  https://twitter.com/dog_rates/status/891815181...  ...  0.078253    True   
3  https://twitter.com/dog_rates/status/891689557...  ...  0.168086    True   
4  https://twitter.com/dog_rates/status/891327558...  ...  0.225770    True   

                            p3   p3_conf p3_dog  \
0                       banana  0.076110  False   
1                     papillon  0.068957   True   
2                       kelpie  0.031379   True   
3                      spatula  0.040836  False   
4  German_short-haired_pointer  0.175219   True   

                                           full_text favorite_count  \
0  This is Phineas. He's a mystical boy. Only eve...        35458.0   
1  This is Tilly. She's just checking pup on you....        30695.0   
2  This is Archie. He is a rare Norwegian Pouncin...        23088.0   
3  This is Darla. She commenced a snooze mid meal...        38741.0   
4  This is Franklin. He would like you to stop ca...        37026.0   

  retweet_count  favorited possibly_sensitive  
0        7487.0      False              False  
1        5557.0      False              False  
2        3680.0      False              False  
3        7664.0      False              False  
4        8265.0      False              False  

[5 rows x 33 columns]

In [103]:
merged2.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo',
       'jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf',
       'p2_dog', 'p3', 'p3_conf', 'p3_dog', 'full_text', 'favorite_count',
       'retweet_count', 'favorited', 'possibly_sensitive'],
      dtype='object')

In [157]:
#copy the raw data so it's still there if we need it, available as merged2


df = merged2.copy()
df.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  ...   p2_conf  p2_dog  \
0  https://twitter.com/dog_rates/status/892420643...  ...  0.085851   False   

       p3  p3_conf p3_dog                                          full_text  \
0  banana  0.07611  False  This is Phineas. He's a mystical boy. Only eve...   

  favorite_count retweet_count  favorited possibly_sensitive  
0        35458.0        7487.0      False              False  

[1 rows x 33 columns]

## Cleaning The Data
Now that we have our data in the correct format, we can go through our issues one by one.

QUALITY:

1. rating Numerator - drop observations outwith (1-20)
2. Rating denominator - drop observations outwith (10)
3. Predictions- drop excess columns
4. Dog Name - lot of a's there - just change to None
5. Text - check that this is the same as full_text
6. Retweeted - "you only want original rating that have images"
6.2 - jpg_uirl
7. Timestamp

8.Possibly Sensitive - in fact all extraneous columns

TIDY:

1. Rating
2. Dog_Status
3. Remove full_text as duplicate


## Q1: rating Numerator 

### Define: drop observations outwith (1-20)


### Code:

In [158]:

df = df.loc[(df.rating_numerator > 0) & (df.rating_numerator < 21)]


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  ...   p2_conf  p2_dog  \
0  https://twitter.com/dog_rates/status/892420643...  ...  0.085851   False   

       p3  p3_conf p3_dog                                          full_text  \
0  banana  0.07611  False  This is Phineas. He's a mystical boy. Only eve...   

  favorite_count retweet_count  favorited possibly_sensitive  
0        35458.0        7487.0      False              False  

[1 rows x 33 columns]

### Test:

In [2]:
df.rating_numerator.value_counts()

NameError: name 'df' is not defined

## Q2 - rating Denominator

### Define:

Drop observations that are not 10

### Code: 

In [ ]:
df = df.loc[(df.rating_denominator == 10)]

### Test:

In [ ]:
df.rating_denominator.value_counts()

## Q3 - Predictions

### Define:

Drop excess columns

### Code:

In [ ]:
excess_columns = ["p2", "p2_conf", "p2_dog", "p3", "p3_conf", "p3_dog"]
df = df.drop(columns=excess_columns)

### Test:

In [ ]:
df.info()

## Q4 - Predictions

### Define:

Change "a"'s to None

### Code:

In [ ]:
df.name.replace(to_replace="a", value = None, inplace=True)

### Test:

In [ ]:
df.name.value_counts()

## Q5 - Text

### Define:

Fix text and full_text columns not matching
Choose one to drop

### Code:


In [162]:
#5 shows that the text column has more data - mostly retweets?
different_text = df[df.text != df.full_text]
different_text[["text", "full_text"]]

text full_text
19    RT @dog_rates: This is Canela. She attempted s...       NaN
95    RT @dog_rates: This is Walter. He won't start ...       NaN
101   RT @loganamnosis: Penelope here is doing me qu...       NaN
104   This is Harry. His ears are activated one at a...       NaN
118   RT @dog_rates: We only rate dogs. This is quit...       NaN
132   RT @dog_rates: This is Jamesy. He gives a kiss...       NaN
155   RT @dog_rates: "Good afternoon class today we'...       NaN
182   RT @dog_rates: This is Luna. It's her first ti...       NaN
211   RT @dog_rates: This is Astrid. She's a guide d...       NaN
247   RT @dog_rates: Here's a heartwarming scene of ...       NaN
253   This is Luna. It's her first time outside and ...       NaN
260   RT @dog_rates: This is Stephan. He just wants ...       NaN
296   This is Lucy. She has a portrait of herself on...       NaN
298   RT @KennyFromDaBlok: 14/10 h*ckin good hats. w...       NaN
363   This is Astrid. She's a guide doggo in trainin...       NaN
382   RT @dog_rates: This is Phil. He's an important...       NaN
506   RT @dog_rates: Meet Sammy. At first I was like...       NaN
566   RT @dog_rates: Everybody drop what you're doin...       NaN
750   This is Reggie. He hugs everyone he meets. 12/...       NaN
815   RT @katieornah: @dog_rates learning a lot at c...       NaN
818   RT @dog_rates: Here's a doggo blowing bubbles....       NaN
890   RT @dog_rates: This... is a Tyrannosaurus rex....       NaN
932   This is Charlie. He pouts until he gets to go ...       NaN
1726  Meet Sammy. At first I was like "that's a snow...       NaN

In [164]:
#5 conclusion - drop full_text as less complete
df.drop(columns="full_text",inplace=True)

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo',
       'jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'favorite_count',
       'retweet_count', 'favorited', 'possibly_sensitive'],
      dtype='object')

### Test:

In [139]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2322 entries, 0 to 2355
Data columns (total 27 columns):
tweet_id                      2322 non-null int64
in_reply_to_status_id         69 non-null float64
in_reply_to_user_id           69 non-null float64
timestamp                     2322 non-null object
source                        2322 non-null object
text                          2322 non-null object
retweeted_status_id           179 non-null float64
retweeted_status_user_id      179 non-null float64
retweeted_status_timestamp    179 non-null object
expanded_urls                 2270 non-null object
rating_numerator              2322 non-null int64
rating_denominator            2322 non-null int64
name                          2322 non-null object
doggo                         2322 non-null object
floofer                       2322 non-null object
pupper                        2322 non-null object
puppo                         2322 non-null object
jpg_url                       20

## Q6 - Retweets

### Define:

Drop observations that are retweets - ie have a value in retweeted_status_id

### Code: 

In [174]:

#6 get rid of retweets - anything that has a value in retweeted_status_id
#is not required
df = df[df.retweeted_status_id.isnull()]


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  ...  puppo  \
0     https://twitter.com/dog_rates/status/892420643...  ...   None   
1     https://twitter.com/dog_rates/status/892177421...  ...   None   
2     https://twitter.com/dog_rates/status/891815181...  ...   None   
3     https://twitter.com/dog_rates/status/

### Test:

In [175]:
df.retweeted_status_id.value_counts()

Series([], Name: retweeted_status_id, dtype: int64)

## Q7 - Timestamp

### Define:

Change timestamp column datatype to datetime object

### Code: 

In [177]:
#7 sort timestamp column
df.timestamp = df.timestamp.astype("datetime64")

### Test:

In [ ]:
df.info()

## Q8 - Removing extra columns

### Define:

Drop columns that do not interest us.

### Code: 

In [179]:
#8 possibly sensitive column doesnt tell us anything

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo',
       'jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'favorite_count',
       'retweet_count', 'favorited', 'possibly_sensitive'],
      dtype='object')

In [180]:
uneeded_columns = ["in_reply_to_status_id", "in_reply_to_user_id", "source", "retweeted_status_id", "retweeted_status_user_id", "retweeted_status_timestamp", "expanded_urls", "jpg_url", "possibly_sensitive"]

df.drop(columns= uneeded_columns,inplace=True)

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...                13   

   rating_denominator     name doggo floofer pupper puppo  img_num      p1  \
0                  10  Phineas  None    None   None  None      1.0  orange   

    p1_conf p1_dog  favorite_count  retweet_count favorited  
0  0.097049  False         35458.0         7487.0     False

### Test:

In [ ]:
df.head(1)

# End of Data Cleaning - Quality.


# On to Data Cleaning - Tidiness

1. Rating - needs to be one colum - drop denominator, rename df.numerator
2. Dog_Status - pandas melt


## T1 - Fix Rating

### Define:
Drop the Denominator Column and rename Numerator column

### Code: 

In [182]:
mapper = {"rating_numerator": "rating_out_of_10"}
df = df.rename(mapper, axis=1)
df.drop("rating_denominator", inplace=True, axis=1)

Index(['tweet_id', 'timestamp', 'text', 'rating_out_of_10',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo',
       'img_num', 'p1', 'p1_conf', 'p1_dog', 'favorite_count', 'retweet_count',
       'favorited'],
      dtype='object')

### Test:

In [184]:

df.head(1)

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   

                                                text  rating_out_of_10  \
0  This is Phineas. He's a mystical boy. Only eve...                13   

      name doggo floofer pupper puppo  img_num      p1   p1_conf p1_dog  \
0  Phineas  None    None   None  None      1.0  orange  0.097049  False   

   favorite_count  retweet_count favorited  
0         35458.0         7487.0     False

## T2 - Fix Dog_status

### Define:
Replace the various pupper, floof etc one hot encoded columns with a single Dog_status column, with these values as categoricals

### Code: 

In [187]:
cols = list(df.columns)
to_remove = ["doggo", "floofer", "pupper", "puppo"]
idvars = [x for x in cols if x not in to_remove]
df.loc[df.doggo == 'doggo', 'dog_class'] = 'doggo'
df.loc[df.floofer == 'floofer', 'dog_class'] = 'floofer'
df.loc[df.pupper == 'pupper', 'dog_class'] = 'pupper'
df.loc[df.puppo == 'puppo', 'dog_class'] = 'puppo'
df.drop(to_remove, inplace=True, axis=1)


['tweet_id',
 'timestamp',
 'text',
 'rating_out_of_10',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo',
 'img_num',
 'p1',
 'p1_conf',
 'p1_dog',
 'favorite_count',
 'retweet_count',
 'favorited']

### Test:


#keeping this as record of mistakes

melted = df.melt(id_vars=idvars,var_name = "doggo_status")
melted.info()
melted.doggo_status.value_counts()

In [200]:
df.p1

0                       orange
1                    Chihuahua
2                    Chihuahua
3                  paper_towel
4                       basset
                 ...          
2351        miniature_pinscher
2352       Rhodesian_ridgeback
2353           German_shepherd
2354                   redbone
2355    Welsh_springer_spaniel
Name: p1, Length: 2143, dtype: object

In [201]:
## Storing and acting on Wrangled Data

df.to_csv('df.csv')